server/src/com/thoughtworks/go/server/websocket/JobInstanceStatusMonitor.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2015 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.server.websocket;
+
+import com.thoughtworks.go.domain.JobInstance;
+import com.thoughtworks.go.domain.JobResult;
+import com.thoughtworks.go.server.domain.JobStatusListener;
+import com.thoughtworks.go.server.service.JobInstanceService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class JobInstanceStatusMonitor {
+
+    @Autowired
+    public JobInstanceStatusMonitor(final AgentRemoteHandler agentRemoteHandler, JobInstanceService jobInstanceService) {
+        jobInstanceService.registerJobStateChangeListener(new JobStatusListener() {
+            @Override
+            public void jobStatusChanged(JobInstance job) {
+                if (job.isRescheduled() || JobResult.Cancelled.equals(job.getResult())) {
+                    agentRemoteHandler.sendCancelMessage(job.getAgentUuid());
```

Comment:
@jyotisingh 
1. We need handle reschedule case for robustness. When agent disconnect and then reconnect, server side may lost track its status, hence server tries to assign new work to it, when we assigning new work to agent, we will check agent status and reschedule job that was assigned to the agent for whatever reason, see [BuildAssignmentService#assignWorkToAgent](https://github.com/gocd/gocd/blob/master/server/src/com/thoughtworks/go/server/service/BuildAssignmentService.java#L107). So listening to reschedule and tell agent cancel current work is necessary, but we don't rely on it, because we will try to cancel running work before we start a new work. I think it's a simple choice to just do it, especially when I don't know how much cases will trigger reschedule, and later we may add new way (or reason) to reschedule a job.
2. Server side will always turn on websocket communication, it's agent's choice to not use websocket right now. When there is no agent connected with websocket, there is almost not cost to loop through JobStateChangeListener